contributors/design-proposals/stable-pods.md

```diff
@@ -0,0 +1,100 @@
+when used by workload controllers.
+3. The workload controllers cannot take advantage of readiness checks when checking the
+state of their underlying Pods.
+4. It's very hard to reason about which pods are not stable just by looking in the status
+of a controller, because it only reports aggregated data. Imagine a ReplicaSet with 200
+replicas, and only 3 of them are unstable.
+
+## Design
+All problems above can be solved by moving MinReadySeconds in the PodSpec. Once a kubelet
+observes that a Pod has been Ready for at least MinReadySeconds without any of its
+containers crashing and passes its readiness check successfully for another time, it will
+update the PodStatus with an Stable Condition set to Status=True. Higher-level
+orchestrators running on different machines such as the ReplicaSet or the
+PodDisruptionBudget controller will merely need to look at the Stable condition that is set
+in the status of a Pod.
+
+Services should not be affected by the implementation of Stable Pods. Endpoints should
+continue to be added or removed when a Pod transitions to or from Ready.
+
+### API changes
+
+A new field is proposed in the PodSpec:
+```go
+	// Minimum number of seconds for which a newly created Pod should be Ready
+	// without any of its container crashing, for it to be considered Stable.
+	// Defaults to 0 (pod will be considered Stable as soon as it is Ready)
+	// +optional
+	MinReadySeconds *int32 `json:"minReadySeconds,omitempty"`
+```
+and a new PodConditionType:
+```go
+	// PodStable is added in a Pod after it has been Ready for at least MinReadySeconds.
+	// The Pod should already have an Endpoint and serve requests, this condition lets
+	// higher-level orchestrators know that the Pod is stable after some amount of time
+	// without having any of its containers crashed.
+	PodStable PodConditionType = "Stable"
+```
+
+The reason for adding a separate Condition instead of extending the existing Ready
+condition is mainly for allowing traffic in Pods. Users can have robust rollouts if their
+applications expose metrics which are then fed into readiness checks.
+
+Additionally:
+* Deployments/ReplicaSets/DaemonSets already use MinReadySeconds in their Spec so we
+should probably deprecate those fields in favor of the field in the PodTemplateSpec and
+remove them in a future version. 
+* All the aforementioned controllers will not propagate MinReadySeconds from their Spec
+down to their PodTemplateSpec because that may lead in differences in the PodTemplateSpec
+between a Deployment and a ReplicaSet resulting in new rollouts. If MinReadySeconds is
+specified both in the Spec and PodTemplateSpec for a controller, the field in the
+PodTemplateSpec will stomp the field in the Spec. If it is specified only in the Spec,
+Pods can be created using it without updating the controller's PodTemplateSpec.
+
+### kubelet changes
+For a Pod that specifies MinReadySeconds, kubelet will need to rerun the readiness check
```

Comment:
My understanding is that ideally that check would be able to identify failures that are not caught by readiness (by using metrics exposed by the app), so we may want to think about introducing a net-new probe. @kubernetes/sig-node-proposals @janetkuo @smarterclayton Or do you guys think re-running the readiness check is enough?